core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -277,7 +367,16 @@ public boolean contains(FieldIndex<T> index, Object value) {
    */
   // TODO(gpang): Remove this method, if it is not being used.
   public Set<T> getByField(FieldIndex<T> index, Object value) {
-    Set<T> set = getByFieldInternal(index, value);
+    Set<T> set;
+    if (index instanceof UniqueFieldIndex) {
```

Comment:
instead of having an `instanceof` to swtich the code path, why not have two overloading methods:  
```java
public Set<T> getByField(NonUniqueFieldIndex<T> index, Object value) 
```
and
```java
public T getByField(UniqueFieldIndex<T> index, Object value) 
```